### PR TITLE
Issue 112 unit based combat

### PR DESCRIPTION
**Pull request by Loomie**

### CODE REVIEW

File: model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/combat/Weapon.kt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+package de.gleex.pltcmd.model.elements.combat
+
+/**
+ * Data that describes what a type of weapon is able to do.
+ * @param roundsPerMinute how many shots this weapon can do per minute including aiming and reloads ([sustained rate](https://en.wikipedia.org/wiki/Rate_of_fire#Sustained_or_effective_rate))
+ * @param shotAccuracy the probability for a single shot to hit the typically aimed for target at a typical range for the weapon category
```
Comment (Baret) [inline model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/combat/Weapon.kt:6]:
```suggestion
 * @param shotAccuracy the probability for a single shot to hit the typically aimed for target at the effective range for the weapon
```

File: model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/combat/Weapon.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package de.gleex.pltcmd.model.elements.combat
+
+/**
+ * Data that describes what a type of weapon is able to do.
```
Comment (Baret) [inline model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/combat/Weapon.kt:4]:
```suggestion
 * Data that describes what a weapon is able to do.
```
The class describes a _weapon_ not a _type of weapon_. And I think for our purposes that is perfectly fine for now.

With weapon _types_ you would need different weapons with that specific type which is one level of hierarchy too much, I believe.

File: model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/units/Units.kt [hunk 3]
```diff
@@ -59,27 +62,27 @@ enum class Units(
...
      * capable of damaging enemy aerial units.
      */
-    AntiAirTeam(personnel = 2),
+    AntiAirTeam(personnel = 2, weapon = Weapons.rpg),
     /**
      * Scouts are specialized in moving through unknown terrain to explore it. They are slower than
      * aerial reconnaissance ([ScoutPlane] for example), but they are much more sneaky.
      */
-    Scout,
+    Scout(weapon = Weapons.pistol),
```
Comment (Baret) [inline model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/units/Units.kt:80]: I think we should not send out any soldiers with nothing but a pistol. Maybe they get an SMG? :)

File: model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/units/Units.kt [hunk 4]
```diff
@@ -204,7 +207,7 @@ enum class Units(
      * The "death from above" dominates the sky. A gunship is mobile and heavily armed so it can provide CAS
      * (close air support) for ground elements no matter what enemy they face.
      */
-    HelicopterGunship(AerialHeavy, personnel = 2, personnelMinimum = 2);
+    HelicopterGunship(AerialHeavy, personnel = 2, personnelMinimum = 2, weapon = Weapons.tankGun);
```
Comment (Baret) [inline model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/units/Units.kt:210]: I really liked the simple but descriptive idea of having `Weapon`s that are carried by `Unit`s... up until I read this line. I already flinched when reading "sniper team" and that they only carry one weapon, but thought it would be ok for simulation purposes.
But here I think it does not quite hit the point. A gunship clearly has a variety of weapons to engage different types of targets and so I thought we rather need some kind of mapping for what targets are engaged with which weapon.
We can keep it as it is for now but later each unit should try to pick the "best" (aka. most valueable or dangerous) target depending on what it can engage. An AT-Soldier will engage armored targets first but when there are none he acts like a normal rifleman.

File: game/engine/src/test/kotlin/de/gleex/pltcmd/game/engine/systems/behaviours/FightingTest.kt [hunk 5]
```diff
@@ -78,6 +78,7 @@ class FightingTest : StringSpec({
 private fun setupCombat(attackerPosition: Coordinate, attackerStats: CombatStats, vararg targetStats: CombatStats): Pair<ElementEntity, GameContext> {
 
     val context = mockk<GameContext>()
+    every { context.random } returns Random(123L)
```
[the remark this answers is not in the document]
Comment (Loomie) [reply game/engine/src/test/kotlin/de/gleex/pltcmd/game/engine/systems/behaviours/FightingTest.kt:81]: That's what mocking frameworks are for 👅

File: game/engine/src/test/kotlin/de/gleex/pltcmd/game/engine/systems/behaviours/FightingTest.kt [hunk 6]
```diff
@@ -5,38 +5,38 @@ import de.gleex.pltcmd.game.engine.entities.types.*
...
-        Fighting.attackNearbyEnemies(attacker, context)
-        assertCombatResult(attackerStats, targetStats, 60, true)
-
-        Fighting.attackNearbyEnemies(attacker, context)
-        assertCombatResult(attackerStats, targetStats, 40, true)
-
-        Fighting.attackNearbyEnemies(attacker, context)
-        assertCombatResult(attackerStats, targetStats, 20, true)
+        var expectedHealth = targetStats.health.value
+        listOf(7, 5, 7, 6, 4, 8, 3, 8, 7, 5, 8, 6, 6, 6, 8, 5).forEach { expectedDamage ->
```
[the remark this answers is not in the document]
Comment (Loomie) [reply game/engine/src/test/kotlin/de/gleex/pltcmd/game/engine/systems/behaviours/FightingTest.kt:34]: Good point.

File: game/engine/src/main/kotlin/de/gleex/pltcmd/game/engine/attributes/CombatAttribute.kt [hunk 7]
```diff
@@ -1,7 +1,12 @@
 package de.gleex.pltcmd.game.engine.attributes
 
+import de.gleex.pltcmd.model.elements.Element
 import de.gleex.pltcmd.model.elements.combat.CombatStats
 import org.hexworks.amethyst.api.Attribute
+import org.hexworks.cobalt.databinding.api.extension.toProperty
 
 /** The [CombatStats] of a fighting entity. */
-internal class CombatAttribute(val stats: CombatStats = CombatStats()) : Attribute
+internal class CombatAttribute(element: Element) : Attribute {
```
[the remark this answers is not in the document]
Comment (Loomie) [reply game/engine/src/main/kotlin/de/gleex/pltcmd/game/engine/attributes/CombatAttribute.kt:9]: That's exactly the point of this class: This attribute has only the single field `CombatStats` instead of all individual stats which might not be needed by everybody. Therein all the data needed for the combat computation is wrapped. It cannot be split up because it contains all data needed for combat. So everything that fights needs this, period. If anything only uses a part of it (maybe movement wants to incorporate the health level) why not. If more abstracted is needed we could infer an interface for that as well. But putting that in two objects would require more attributes to be combined together and calls get a long list of parameters. OK, maybe we can use composition for this if needed: The CombatStats may get a reference to the HealthStats which is also used by other attributes.
I will keep this in mind to extract data to own classes if needed.

> Currently I can only think of maybe offensive and defensive combat stats. 
Defensive stats are not yet implemented. I wanted to keep it simple for the first run ;)

File: model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/combat/CombatStats.kt [hunk 8]
```diff
@@ -5,23 +5,29 @@ import org.hexworks.cobalt.databinding.api.extension.toProperty
...
         get() = alive.value
 
-    fun attack(target: CombatStats) {
+    fun attack(target: CombatStats, random: Random) {
         if (target.isAlive) {
-            target.health.transformValue { it - firepower.value }
-            log.debug("attack with ${firepower.value} fire power resulted in target health of ${target.health.value}")
+            val damagePerTick = weapons.value.map { it.fireShots(random) }
+                    .sum()
+            val receivedDamage = min(damagePerTick, target.health.value)
```
[the remark this answers is not in the document]
Comment (Loomie) [reply model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/combat/CombatStats.kt:28]: Because at zero health points you are dead. If you get more damage you cannot get more dead (deadier? there is not even a word for in english). But for your idea about special effects it might be nice. But we could track that separately. Maybe a DamageListener can inform interested parties about the real number. But for tracking the liveness this is not relevant (and therefore could also be negative).
But the inital unit test wanted to assert that the health is zero ;)

File: model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/combat/CombatStats.kt [hunk 9]
```diff
@@ -37,4 +43,16 @@ data class CombatStats(val firepower: ObservableValue<Int> = 20.toProperty(), va
         }
     }
 
-}
\ No newline at end of file
+    /** @return damage of hits per tick */
+    private fun Weapon.fireShots(random: Random): Int {
+        val shotsPerTick = roundsPerMinute / 1 // TODO make GameConstants.Time.ticksPerMinute available
```
[the remark this answers is not in the document]
Comment (Loomie) [reply model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/combat/CombatStats.kt:48]: Yeah `model` sounds good. I don't think that such core game concepts are utils.

File: game/engine/src/main/kotlin/de/gleex/pltcmd/game/engine/attributes/CombatAttribute.kt [hunk 10]
```diff
@@ -1,7 +1,12 @@
...
+    val stats: CombatStats = CombatStats(element.units.map { it.blueprint.weapon }
```
[the remark this answers is not in the document]
Comment (Loomie) [reply game/engine/src/main/kotlin/de/gleex/pltcmd/game/engine/attributes/CombatAttribute.kt:10]: I think this is correct as it is. Only the office and his closest mates can shoot from their position. If they command sub elements somewhere else on the map (with an own position), these other elements cannot attack the same target.

Hm, so I don't know if we can always say: Hey brigade commander attack with you whole army the soldier next to you. Or if each element should attack on its own. So from my perspective each elements needs to get its own attack command and therefor call to its own attributes. As the issue says: I built unit based combat.

File: model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/combat/CombatStats.kt [hunk 11]
```diff
@@ -37,4 +43,16 @@ data class CombatStats(val firepower: ObservableValue<Int> = 20.toProperty(), va
...
+        var hits = 0
+        repeat(shotsPerTick) {
+            if (random.nextDouble() <= shotAccuracy) {
+                hits = hits.inc()
+            }
+        }
+        log.trace("firing $shotsPerTick shots with $shotAccuracy each results in $hits hits")
+        return hits * 1 // dmg / shot
```
[the remark this answers is not in the document]
Comment (Loomie) [reply model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/combat/CombatStats.kt:75]: https://github.com/Baret/pltcmd/issues/115

File: model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/combat/Weapon.kt (hunk 1 above)
Comment (Loomie) [reply model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/combat/Weapon.kt:6]: No. I wanted to compare uneven weapons with the same stat: A pistol which has a reach of a few meters and a sniper rifle with over one kilometer range are in sense of stats both hand guns. So to be able to use the same stat for different weapons I thought about categories, because it is hard to use a single number for everything: Hand guns, vehicle weapons, fire support.
I don't know if we need an own model for that or just more values for a weapon.

File: model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/units/Units.kt (hunk 4 above)
Comment (Loomie) [reply model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/units/Units.kt:210]: You are right. Also see my comment about abstracting weapon stats to some numbers: https://github.com/Baret/pltcmd/pull/114/files/5786be934dacb87a90c899c70a33749397c84275#r484061223

File: model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/combat/CombatStats.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (Baret) [reply model/elements/src/main/kotlin/de/gleex/pltcmd/model/elements/combat/CombatStats.kt:40]: > But the inital unit test asserts that the health is zero ;)

In my opinion the assertion is wrong then. The condition for being alive is "hp > 0" which in turn means that you are dead when the hp is <= 0. The assertion says "you are dead when you have exactly 0 hp" but what are you when you have -5 hp?

For now it doesnt' matter but I think the condtion is wrong. "alive" is a boolean represented by "> 0" and "<= 0". When we use "> 0" and "== 0" there is something missing.

File: game/engine/src/main/kotlin/de/gleex/pltcmd/game/engine/attributes/CombatAttribute.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (Baret) [reply game/engine/src/main/kotlin/de/gleex/pltcmd/game/engine/attributes/CombatAttribute.kt:10]: > If they command sub elements somewhere else on the map (with an own position)

The thing is: These lements _don't have a position_! They are not excusively simulated, they are not an entity themselves, they do not have a position attribute.

The player is a commander with a rather abstract look on the battlefield. We only _have to_ simulate that abstraction level for the game to work. But we _may_ do much more detailed simulation and calculations.
For the game to work we might simply add plain attributes to the `CommandingElement`-entity like "morale = 85, speed = 3.5, damage = 30, hp = 57, position = (153|447)". But as the element is made up of units (more than only the commanding element's units) we can aggregate the element's stats from those units. BUT: We currently only have _one_ position for the whole element which means all units are assumed to be in the same coordinate. And this is perfectly fine for now and I think even a long time. That only becomes a problem (or rather falsifies simulation) for really big elements that most probably do not fit into one 100*100m square.

I don't know if I can express what I mean, probably we should have a quick talk about that.

TL;DR version: `Element` is not relevant in the engine, all entities should be based on `CommandingElement`s.

**Comment by Loomie** [comment]
I moved the state for the combat to the engine directly as attributes. The logic has moved to the CombatEntity.

**Review by Baret — APPROVED** [review]
Maybe have some kdoc and then it's fine!
